### PR TITLE
CIRC-222 Send loan-related notices on check-out and check-in

### PR DESCRIPTION
**Pull request by dmytrotur**
## Purpose
Resolves: CIRC-222
Implement sending patron notices on check-in and check-out
## Links
https://issues.folio.org/browse/CIRC-222

### CODE REVIEW

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 1]
```diff
@@ -605,6 +616,68 @@ public void cannotCheckOutWhenServicePointOfCheckoutNotPresent()
       hasServicePointParameter(null))));
   }
 
+  @Test
+  public void patronNoticeOnCheckoutHasBeenSendWhenCheckoutLoanNoticeIsDefined()
```
Comment (marcjohnson-kint) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java]:
```suggestion
  public void checkoutNoticeIsSentWhenPolicyDefinesCheckoutNotice()
```

File: src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  public boolean isRecurring() {
+    return recurring;
+  }
+
+  public Period getRecurringPeriod() {
+    return recurringPeriod;
+  }
+
+
+  public static class Builder {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java]: Is this builder used by production code or only tests?

File: descriptors/ModuleDescriptor-template.json [hunk 3]
```diff
@@ -80,7 +81,8 @@
             "circulation-storage.loan-policies.item.get",
             "circulation-storage.request-policies.item.get",
             "circulation-storage.fixed-due-date-schedules.item.get",
-            "circulation-storage.fixed-due-date-schedules.collection.get"
+            "circulation-storage.fixed-due-date-schedules.collection.get",
+            "patron-notice.post"
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: Please add a dependency for this interface as well

File: src/main/java/org/folio/circulation/domain/LoanRepository.java [hunk 4]
```diff
@@ -51,11 +51,12 @@ public LoanRepository(Clients clients) {
       storageLoan.put("loanPolicyId", loanAndRelatedRecords.getLoanPolicy().getId());
     }
 
+    User user = loanAndRelatedRecords.getLoan().getUser();
     return loansStorageClient.post(storageLoan).thenApply(response -> {
       if (response.getStatusCode() == 201) {
         return succeeded(
           loanAndRelatedRecords.withLoan(Loan.from(response.getJson(),
-            loanAndRelatedRecords.getLoan().getItem())));
+            loanAndRelatedRecords.getLoan().getItem()).withUser(user)));
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/LoanRepository.java]: Can you help me understand this change, why does the user need to be updated?

File: src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java [hunk 5]
```diff
@@ -102,4 +116,23 @@
...
+        return HttpResult.succeeded(records);
+      }));
+  }
+
+  private void sendCheckOutPatronNoticeWhenPolicyFound(CheckInProcessRecords records, PatronNoticePolicy patronNoticePolicy) {
+    Loan loan = records.getLoan();
+    List<NoticeDescriptor> noticeDescriptors =
+      patronNoticePolicy.lookupLoanNoticeDescriptor(NoticeEventType.CHECK_IN, NoticeTiming.UPON_AT);
+    JsonObject noticeContext = patronNoticeService.createNoticeContextFromLoan(loan);
+    patronNoticeService.sendPatronNotice(noticeDescriptors, loan.getUserId(), noticeContext);
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java]: Can there be multiple notices (in the same policy) for a check in?

File: src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java (hunk 2 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java]: Production only

File: src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java [hunk 6]
```diff
@@ -102,4 +116,23 @@
 
     return completedFuture(succeeded(item));
   }
+
+  CompletableFuture<HttpResult<CheckInProcessRecords>> sendCheckInPatronNotice(CheckInProcessRecords records) {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java]: The name of this method refers to a check in notice, yet the invocation within it, is for check out:
`sendCheckOutPatronNoticeWhenPolicyFound(records, policy)`

Which type of notice should this code be sending?

File: src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java [hunk 7]
```diff
@@ -102,4 +116,23 @@
...
+      return CompletableFuture.completedFuture(HttpResult.succeeded(records));
+    }
+    return patronNoticePolicyRepository.lookupNoticePolicy(records.getLoan())
+      .thenApply(r -> r.next(policy -> {
+        sendCheckOutPatronNoticeWhenPolicyFound(records, policy);
+        return HttpResult.succeeded(records);
+      }));
+  }
+
+  private void sendCheckOutPatronNoticeWhenPolicyFound(CheckInProcessRecords records, PatronNoticePolicy patronNoticePolicy) {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java]: This method states it is sending a check out notice, yet uses the check in type?
`patronNoticePolicy.lookupLoanNoticeDescriptor(NoticeEventType.CHECK_IN, NoticeTiming.UPON_AT);`

File: descriptors/ModuleDescriptor-template.json (hunk 3 above)
Comment (dmytrotur) [reply descriptors/ModuleDescriptor-template.json]: Notices are send without waiting for response.
Do you think mod-circulation now has hard dependency on mod-notify?

File: src/main/java/org/folio/circulation/domain/LoanRepository.java (hunk 4 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/LoanRepository.java]: When loan is saved, the **loan** field in **LoanAndRelatedRecords**  is updated and does not contain user representation. This representation is required to send patron notice just after loan is saved successfully.

File: src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java (hunk 5 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java]: Yes, the schema for patron notice policy supports it.

File: src/main/java/org/folio/circulation/support/Clients.java [hunk 8]
```diff
@@ -247,5 +260,19 @@ private CollectionResourceClient createCalendarStorageClient(
     throws MalformedURLException {
     return getCollectionResourceClient(client, context, "/calendar/periods");
   }
-  
+
+  private CollectionResourceClient createPatronNoticePolicesStorageClient(
+    OkapiHttpClient client,
+    WebContext context)
+    throws MalformedURLException {
+    return getCollectionResourceClient(client, context, "/patron-notice-policy-storage/patron-notice-policies");
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/support/Clients.java]: Personally, I'd split this into two lines before the path, as otherwise it is quite a long line

File: src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java (hunk 6 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java]: The idea of this method is to move processing of found patron notice into separate place.
Probably should consider better naming.

File: src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java (hunk 7 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java]: Misspelling, sorry. Should be CheckIn.

File: descriptors/ModuleDescriptor-template.json (hunk 3 above)
Comment (marcjohnson-kint) [reply descriptors/ModuleDescriptor-template.json]: @dmytrotur As far as I am aware, FOLIO only has explicit or implicit dependencies.

Until we have a way of describing optional dependencies, I advocate for explicit dependencies, even if that extends the dependency graph and impacts environment building.

Otherwise it is not possible to tell if the dependency is optional or missing because of a mistake.

Does that make sense?

Are you proposing having an implicit dependency?

File: src/main/java/org/folio/circulation/support/Clients.java (hunk 8 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/support/Clients.java]: Agree, I'll split this.

File: src/test/java/api/loans/CheckOutByBarcodeTests.java (hunk 1 above)
Comment (dmytrotur) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java]: Ok, thanks, I'll change it.

File: src/main/java/org/folio/circulation/domain/LoanRepository.java (hunk 4 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/LoanRepository.java]: `Hmm, LoanAndRelatedRecords.withLoan(...` should not lose the user information that was there, I'd prefer this was fixed internally rather than here.

That might mean checking that the user for the loan is the same, and then keeping the existing user.

If that doesn't make sense, I can try to demonstrate it tomorrow.

File: src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java (hunk 2 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java]: Personally, I'd prefer if it was split into a separate file and named appropriately. I will leave that decision up to you.

File: src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java (hunk 5 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java]: Fair enough, that seems counter intuitive to me, maybe that is a follow up question for Darcy around notice policies?

File: descriptors/ModuleDescriptor-template.json (hunk 3 above)
Comment (dmytrotur) [reply descriptors/ModuleDescriptor-template.json]: Yes, make sense for me. I'll add explicit dependency.

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 9]
```diff
@@ -45,11 +50,17 @@
...
+import api.support.builders.NoticeDescriptorBuilder;
+import api.support.builders.NoticePolicyBuilder;
 import api.support.builders.UserBuilder;
 import io.vertx.core.json.JsonArray;
 import io.vertx.core.json.JsonObject;
 
 public class CheckOutByBarcodeTests extends APITests {
+
+  private static final DateTimeFormatter NOTICE_DATE_TIME_FORMATTER =
+    DateTimeFormat.forPattern("MM/dd/yyyy hh:mm a");
```
Comment (marcjohnson-kint) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java]: Where does this date format come from?

File: src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
+package org.folio.circulation.domain.notice;
+
+import org.joda.time.Period;
+
+public class NoticeDescriptor {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java]: What is a notice descriptor?

File: src/test/java/api/loans/CheckOutByBarcodeTests.java (hunk 9 above)
Comment (dmytrotur) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java]: There is no requirements on how due date should be displayed in patron notice. I've added this as stub.  Do you think we should something default like rfc3339 for now?

File: src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java (hunk 6 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java]: @dmytrotur I'm happy with the separation. I just didn't understand the mismatch between check in and check out within it.

File: src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java (hunk 10 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java]: A record in notice policy describing how notice should be sent for specific event.

File: src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java (hunk 6 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java]: Oh, sorry, it's just misspelling, I'll fix it.

File: src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java (hunk 10 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java]: How does Darcy refer to this, is it using this name?

I'm trying to understand if there is a domain terminology for this.

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java]: Judging from what we have on UI, it is called just 'Notice'. 
image

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java]: Can be called something like 'NoticeSetting'. I'm not really sure about the naming.

File: src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java (hunk 5 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java]: I think so, now there is no validation for that.

File: src/main/java/org/folio/circulation/domain/LoanAndRelatedRecords.java [hunk 11]
```diff
@@ -9,69 +8,57 @@
   private final Loan loan;
   private final RequestQueue requestQueue;
   private final LoanPolicy loanPolicy;
-  private final AdjacentOpeningDays adjacentOpeningDays;
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/circulation/domain/LoanAndRelatedRecords.java]: Why is this being removed with this PR? I think code clean up is great, but, when it is unrelated to the story, I think it should be handled via a separate PR/story.

File: src/main/java/org/folio/circulation/domain/LoanRepository.java (hunk 4 above)
Comment (mreno-EBSCO) [reply src/main/java/org/folio/circulation/domain/LoanRepository.java]: I'm not sure there is anything to fix here. I think the problem is that the loan is being recreated via the response of the `post` to loans-storage. That response will only have IDs. Maybe we should look up the user ID again or at least verify that the IDs match before we jam it into the loan.

If we are going to drop in the user, then we should probably just switch to this to build the loan object:
https://github.com/folio-org/mod-circulation/blob/a720b4667a6c9d9fa86accbc889025fefcbc37e0/src/main/java/org/folio/circulation/domain/Loan.java#L84-L88

File: src/main/java/org/folio/circulation/support/Clients.java [hunk 12]
```diff
@@ -247,5 +260,20 @@ private CollectionResourceClient createCalendarStorageClient(
...
     return getCollectionResourceClient(client, context, "/calendar/periods");
   }
-  
+
+  private CollectionResourceClient createPatronNoticePolicesStorageClient(
+    OkapiHttpClient client,
+    WebContext context)
+    throws MalformedURLException {
+    return getCollectionResourceClient(client, context,
+      "/patron-notice-policy-storage/patron-notice-policies");
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/circulation/support/Clients.java]: Probably not a question for this PR, but why are patron notice policies in mod-circulation-storage? It doesn't really make sense to me, especially with fees/fines data in the policy. This seems like something that should be in its own module. These microservices are starting to become monolithic in my mind. Maybe I am missing something. Sorry for the rant... 😞

File: src/test/java/api/loans/CheckInByBarcodeTests.java [hunk 13]
```diff
@@ -297,4 +305,116 @@ public void canCheckInAnItemTwice()
     assertThat("barcode is included for item",
       itemFromResponse.getString("barcode"), is("565578437802"));
   }
+
+  @Test
+  public void patronNoticeOnCheckInHasBeenSendWhenCheckInLoanNoticeIsDefinedAndLoanExists()
```
Comment (mreno-EBSCO) [inline src/test/java/api/loans/CheckInByBarcodeTests.java]: Maybe: `patronNoticeOnCheckInIsSentWhenCheckInLoanNoticeIsDefinedAndLoanExists`

File: src/test/java/api/loans/CheckOutByBarcodeTests.java (hunk 9 above)
Comment (marcjohnson-kint) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java]: If we are intending to present this in the notice then I think we need to clarify the requirements with Darcy, otherwise this is likely to be rejected during testing.

Please can you ask her what format she wants to use?

Why is this being decided here, rather than by the templating part of the system?

File: src/main/java/org/folio/circulation/domain/LoanRepository.java (hunk 4 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/LoanRepository.java]: @mreno-EBSCO Thanks, yeah, I reckon using the extended from might be clearer.

It’s a minor point, I wondered if it might be clearer to have a dedicated method, to do the remembering of the item and user automatically.

Part of this is the compromises mod-circulation makes in how it models the domain, both the related records (mostly to carry state between async methods) and the retention of JSON representations (to avoid conversions).

File: src/main/java/org/folio/circulation/support/Clients.java (hunk 12 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/support/Clients.java]: @mreno-EBSCO I don’t think you are missing something. 

I think the modules have grown, partly due to the overhead of introducing new modules and partly due to behaviour that can only be accomplished when interfaces are provided by the same module (e.g. constraints in storage).

This is something that likely needs discussing in a more general sense. It may be worth raising a JIRA issue to try to raise awareness of examples like this, and decide upon a policy, of when to move interface implementations to new modules. What do you think?

I think that mod-circulation knowing about the policies is itself a compromise. I’d prefer that mod-circulation was responsible for identifying and publishing the checked in (to return) and checked out (both deliberately past tense) events and a different module was responsible for determining if a policy meant that a notification should be sent.

However I don’t think we have the infrastructure in place for a design like that at the moment.

File: src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java (hunk 10 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java]: I’d prefer something like `NoticeSetting` or `NoticeConfiguration`.

I might even suggest notice if that is what it is called in the domain, however that is likely ambiguous with the actual notice itself. So maybe that is a conversation for another day.

**Comment by dmytrotur** [comment]
@marcjohnson-kint Resolved code duplicates issue by moving common code from LoanPolicyRepository and PatronNoticePolicyRepository to base class CirculationPolicyRepository.

File: src/test/java/api/loans/CheckInByBarcodeTests.java (hunk 13 above)
Comment (dmytrotur) [reply src/test/java/api/loans/CheckInByBarcodeTests.java]: Done

File: src/main/java/org/folio/circulation/domain/LoanAndRelatedRecords.java (hunk 11 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/LoanAndRelatedRecords.java]: Agree, reverted these changes.

File: src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java (hunk 10 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/notice/NoticeDescriptor.java]: Renamed to NoticeConfiguration

File: src/main/java/org/folio/circulation/domain/notice/NoticeDescriptorBuilder.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
+package org.folio.circulation.domain.notice;
+
+import org.joda.time.Period;
+
+public class NoticeDescriptorBuilder {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/notice/NoticeDescriptorBuilder.java]: This needs renaming now the class it creates has been renamed

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/notice/NoticeDescriptorBuilder.java]: Done

**Comment by dmytrotur** [comment]
@marcjohnson-kint Added dependency and required permissions

File: src/main/java/org/folio/circulation/support/Clients.java (hunk 12 above)
Comment (mreno-EBSCO) [reply src/main/java/org/folio/circulation/support/Clients.java]: @marcjohnson-kint I hear you. I've been ranting about this in the past, but there is just too much to do to get basic v1 functionality working that this type of compromise gets swept under the rug, which I totally get.

I think what adds some level of confusion is that, to me, it is not obvious that `patron-notice-policy-storage` would be in mod-circulation-storage. I had to do a folio.org GitHub search to find it. The raml file in mod-circulation-storage doesn't include "-storage" (probably an oversight) in the name, so that is confusing as well.

Anyway, I think we need a coordinated service that can manage internal out-of-band operations across all modules. How we achieve this is probably best left until after v1 or until necessity dictates. Much to my chagrin, this is probably not that case.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/support/Clients.java]: > it is not obvious that patron-notice-policy-storage would be in mod-circulation-storage

Indeed, that is partially deliberate in that the interface name shouldn't reveal the implementor, but it does make finding them harder.

> The raml file in mod-circulation-storage doesn't include "-storage" (probably an oversight) in the name, so that is confusing as well.

That is likely a mistake, that could do with fixing, though it might impact our documentation generation scripts. And there are quite a lot of minor issues like this with RAML and JSON files across the repositories.

> I think we need a coordinated service that can manage internal out-of-band operations across all modules. How we achieve this is probably best left until after v1 or until necessity dictates.

By out of band do you mean some kind of publish / subscribe mechanism and/or something based upon messaging rather than HTTP?

If you do, then I believe there have been some discussions about pub / sub as I believe some work has started on a capability like that. I agree, exploring this topic more is definitely a post 2019 Q1 conversation.

Comment (mreno-EBSCO) [reply src/main/java/org/folio/circulation/support/Clients.java]: By out of band, I meant something not triggered by a REST API call, like on a schedule. Though, this service could be fed by a pub/sub queue, like Kafka. Or use reddis... or workflow tasks... or whatever. 😄 I've been part of the pub/sub discussions and have worked with the AES module. We've got a long way to go, but we are making some progress. Agreed, this is all post Q1 and possibly post v1.

File: src/test/java/api/loans/CheckOutByBarcodeTests.java (hunk 9 above)
Comment (dmytrotur) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java]: I've asked about the format in UICIRC-154 and the answer is to use local date format.
Seems like now it's out of scope for this story.
And the decision where to apply the locale could be taken later.
Also I should mention that the template engine now does not support date time formatting (only strings can be used to fill placeholders in templates).   
What do you think, can we leave this as stub now?

Comment (marcjohnson-kint) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java]: @dmytrotur Please can you raise a story for making this decision and designing a way to do it in a way that respects the chosen locale.

Personally, I would've preferred to use a less opinionated placeholder, that made it obvious it was such and could be acceptable irrespective of locale, e.g. RFC 3339 like you suggested above.

I also believe this really should be the responsibility of the templating system, rather than the system asking to send a templated notice.

I think we are out of time for the 2019 Q1 release to do this, as this work in the current form is intended to be included.

Comment (dmytrotur) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java]: @marcjohnson-kint Ok, I'll raise this story.

Comment (dmytrotur) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java]: Should I change date format to 'RFC 3339'? Or are we going to merge this as is?

Comment (marcjohnson-kint) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java]: @dmytrotur Personally, if it is a small change, I think that would be good.

It makes it explicit that this is potentially temporary. And, at least in theory, could be acceptable across locale choices.

Feel free to check with @kostyantyn-kh if needed.

Comment (dmytrotur) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java]: @marcjohnson-kint I agree, I'll do this.
